### ISSUE TITLE
MD051: False positive for links to headings containing emoji with variation selectors (U+FE0F)

### ISSUE THREAD

**Issue by latipun7** [DavidAnson/markdownlint#2171]
## What did you do?

Linted a markdown file containing emoji headings with in-page links using `npx --yes markdownlint-cli2@0.23.1 '**/*.md'`.

Minimal reproduction — a file `test.md` with this content:

```markdown
## 🛠️ Development

[Link](#-development)
```

## What did you expect to happen?

No errors. On GitHub, the heading `## 🛠️ Development` produces fragment `#-development` (emoji and variation selector U+FE0F are stripped). The link `[Link](#-development)` is valid.

## What actually happened?

`MD051/link-fragments Link fragments should be valid` is reported on the link.

## What version were you using?

markdownlint v0.41.1 (via markdownlint-cli2 v0.23.1)

## What operating system were you using?

Ubuntu 24.04 (WSL2 on Windows 11)

## Additional context

<img width="1171" height="131" alt="Image" src="https://github.com/user-attachments/assets/e19f2db5-3d76-47f1-ba2e-4da6e824bd94" />

It appears markdownlint's fragment normalization encodes the variation selector (e.g. `%EF%B8%8F`) rather than stripping it, producing an expected fragment like `#%EF%B8%8F-development` that does not match the user-provided `#-development`.

GitHub's slug algorithm treats emoji + variation selector sequences as non-lexical and removes them entirely during fragment generation. This affects any heading containing emoji with variation selectors, extremely common for 👍, 🛠️, ⚡, 📦, 🎉, etc.

**Comment by DavidAnson** [comment]
If you have a link to the specification for the GitHub behavior, that would be helpful. Thank you!

**Comment by DavidAnson** [comment]
You seem to have this backwards. The link GitHub produces for your example is `#%EF%B8%8F-development` and using that with `markdownlint` produces no issues:

https://dlaa.me/markdownlint/#%25m%23%20%F0%9F%9B%A0%EF%B8%8F%20Development%0A%0A%5BDevelopment%5D(%23%25EF%25B8%258F-development)%0A

If there is an example page on GitHub that behaves as you say, please link to it and I'll have a look.

**Comment by DavidAnson** [comment]
For example: https://github.com/DavidAnson/markdownlint/blob/i2171/test/github-auto-identifiers.md#%EF%B8%8F-development

**Comment by latipun7** [comment]
OMG!
I'm sorry, you are right, I thought leading emojis always be stripped because of my browser. I never see the full URL because my browser only show domain if the address bar not focused. Furthermore, I always see the full URL from hover 🤦

<img width="612" height="93" alt="Image" src="https://github.com/user-attachments/assets/29fafe65-28f9-428e-ae73-6f7e1d92058b" />

Turned out my browser give me stripped false full URL on hover, but after clicking it, in address bar give correct URL.

Because of that, I assumed GitHub would completely strip all emojis along with their hidden variation selectors! It turned out GitHub actually strips the base emoji character in both cases, but the resulting link either keeps a hidden variation selector (`## 🛠️ Development` → `#%EF%B8%8F-development`) or leaves nothing behind but the space, which turns into a leading dash (`## 📦 Release Artifacts` → `#-release-artifacts`).

Sorry for the false issue report, thank you for your support 🙏
